### ISSUE TITLE
Loading quantized models with SentenceTransformers 

### ISSUE THREAD

**Issue by emanjavacas** [UKPLab/sentence-transformers#2643]
Dear maintainers,

Does anybody know if loading quantized model is possible with `sentence_transformers`? I am currently looking at embedding models, and some of them like Qwen 7B seem to be `SentenceTransformer` models: https://huggingface.co/Alibaba-NLP/gte-Qwen1.5-7B-instruct

However, `SentenceTransformer` loading code only accepts the model name, and doesn't expose the underlying transformers loading logic. For this reason, I can't find how to do the quantization (for example using `quanto`) before loading: https://huggingface.co/docs/transformers/en/quantization

I can't see why this wouldn't be compatible, since SentenceTransformers uses Transformers anyway under the hood.

Any hints appreciated!

**Comment by tomaarsen** [comment]
Hello!

We're currently investigating the best approach to add this support in #2578. In particular, that PR will expose some parameters (`model_kwargs`, `config_kwargs`, `tokenizer_kwargs`) to the `SentenceTransformer` class to allow e.g. easy quantization. Until then, the best solution is to load the `Transformer` class separately and use the existing `model_args`. However, I'm not 100% sure if that'll work, as you also end up passing the `quantization_config` to the AutoConfig.

Something like:
```python
from sentence_transformers.models import Transformer, Pooling, Normalize
from sentence_transformers import SentenceTransformer

transformer = Transformer("Alibaba-NLP/gte-Qwen1.5-7B-instruct", model_args={"trust_remote_code": True, "quantization_config": ...})
pooling = Pooling(transformer.get_word_embedding_dimension(), pooling_mode="lasttoken")
normalize = Normalize()
model = SentenceTransformer(modules=[transformer, pooling, normalize])
```
(Untested!)

- Tom Aarsen

**Comment by emanjavacas** [comment]
Well, there's only one way to know. I am gonna try and report.
Thanks!

**Comment by emanjavacas** [comment]
I was able to run the model, which is cool and thanks a lot for the hints.

**Comment by alamgirqazi** [comment]
This worked for me. 

```
    def __init__(self, model_name: str, quantize: bool = False):
      
        self.quantize = quantize

        if self.quantize:
            quantization_config = BitsAndBytesConfig(
                load_in_8bit=True,  
                llm_int8_threshold=6.0 
            )

            transformer = Transformer(
                model_name_or_path=model_name,
                model_args={
                    "trust_remote_code": True,
                    "quantization_config": quantization_config,
                    "device_map": "auto"  
                }
            )
            print('quantized')
        else:
            print('not quantized')
            # Initialize the Transformer without quantization
            transformer = Transformer(
                model_name_or_path=model_name,
                model_args={"trust_remote_code": True,
                            }
            )

        pooling = Pooling(transformer.get_word_embedding_dimension(), pooling_mode="mean")
        normalize = Normalize()

        # Create the SentenceTransformer model
        self.model = SentenceTransformer(modules=[transformer, pooling, normalize])

```

**Comment by jchwenger** [comment]
Hi @tomaarsen,
I'm encountering this issue as well. I'm on a Mac M2 with 24GB of memory, and my hope was to be able to test a model too large to fit in memory (e.g. the Qwen3-Embedding-8B, or even the 4B, which runs but very slowly) in quantized form. However, I don't seem to find out how to load an already quantized gguf file (`SentenceTransformer("../models/Qwen/Qwen3-Embedding-8B-Q8_0.gguf")`), or even better, being able to specify which quantized file to choose from in this or that repository (like [here](https://huggingface.co/Qwen/Qwen3-Embedding-8B-GGUF)). I'm guessing this is not meant to be used in that way? Thanks in advance!

**Comment by tomaarsen** [comment]
@jchwenger I'm afraid GGUF models can only be loaded and used by certain projects like llama.cpp ([see docs for the Qwen3-Embedding-8B model](https://huggingface.co/Qwen/Qwen3-Embedding-8B-GGUF#llamacpp)). With Sentence Transformers, you have to load a model based on torch, ONNX, or OpenVINO instead, like the non-GGUF [Qwen3-Embedding-8B](https://huggingface.co/Qwen/Qwen3-Embedding-8B).

- Tom Aarsen

**Comment by jchwenger** [comment]
Thanks @tomaarsen for the clarification! Indeed, somehow I assumed the same logic applied here. Which means if I want a specific quantization that works with my hardware, I need to do it myself using code examples as above (in a CUDA-supporting system, I seem to recall `accelerate` and `quanto` require CUDA), or find one that's already done (like [here](https://huggingface.co/models?other=base_model:quantized:Qwen/Qwen3-Embedding-8B)). Sounds good!

**Comment by Davidyz** [comment]
> I'm afraid GGUF models can only be loaded and used by certain projects like llama.cpp

Hi @tomaarsen , regarding the gguf issue, the `transformers` library [provides gguf support](https://huggingface.co/docs/transformers/en/gguf) (as a kwarg of the `from_pretrain` class method). Since sentence transformers use `transformers` when using the `torch` backend, maybe it's feasible to support gguf for `torch` backend as a starting point?

**Comment by tomaarsen** [comment]
@Davidyz That is pretty interesting actually, would be pretty interesting! The `from_pretrained` kwargs from `transformers` are accessible with `SentenceTransformers` via the `model_kwargs` argument, so perhaps it should already work out of the box.

I think one of the remaining issues is that the other files are not preserved in most GGUF repositories, e.g.: https://huggingface.co/Casual-Autopsy/snowflake-arctic-embed-l-v2.0-gguf/tree/main
Meaning that Sentence Transformers will not be able to recognize e.g. which modules to run.

- Tom Aarsen

**Comment by Davidyz** [comment]
> The `from_pretrained` kwargs from `transformers` are accessible with `SentenceTransformers` via the `model_kwargs` argument, so perhaps it should already work out of the box.

@tomaarsen turned out this is correct. gguf models _almost_ works out of the box. The only things I needed were [gguf](https://pypi.org/project/gguf/) and [accelerate](https://github.com/huggingface/accelerate). After installing these 2 packages, I was able to load and run a [qwen3-embedding gguf](https://huggingface.co/yomir/Qwen3-Embedding-0.6B-GGUF) by setting `model_kwargs["gguf_file"]` to the path to the downloaded gguf file. Loading models from a gguf-only repo doesn't work because `transformers` requires a `config.json` to read `model_type` from. In summary, we just need to set `model_name` to the "upstream" repo (wherever there's a valid `config.json` for the model, like `Qwen/Qwen3-Embedding-0.6B`), and have the gguf downloaded to a local directory and put the path to `model_kwagrs["gguf_file"]`:
```python
model = SentenceTransformer(
    "Qwen/Qwen3-Embedding-0.6B",
    model_kwargs={
        "gguf_file": '/mnt/Shelf/hf_home/Qwen3-Embedding-0.6B-Q8_0.gguf'
    }
)
```
This is probably too sketchy to be included in sentence transformers, but maybe this can be documented somewhere so that it's easier for people who need to run quants to find the instructions.

EDIT: you could probably also just place the `config.json` from the upstream hf repo into the same dir as the gguf, and use that directory as `model_name`. Haven't tested this though.

**Comment by mspinelli** [comment]
> [@tomaarsen](https://github.com/tomaarsen) turned out this is correct. gguf models _almost_ works out of the box. The only things I needed were [gguf](https://pypi.org/project/gguf/) and [accelerate](https://github.com/huggingface/accelerate). After installing these 2 packages, I was able to load and run a [qwen3-embedding gguf](https://huggingface.co/yomir/Qwen3-Embedding-0.6B-GGUF) by setting `model_kwargs["gguf_file"]` to the path to the downloaded gguf file. Loading models from a gguf-only repo doesn't work because `transformers` requires a `config.json` to read `model_type` from. In summary, we just need to set `model_name` to the "upstream" repo (wherever there's a valid `config.json` for the model, like `Qwen/Qwen3-Embedding-0.6B`), and have the gguf downloaded to a local directory and put the path to `model_kwagrs["gguf_file"]`:
> 
> model = SentenceTransformer(
>     "Qwen/Qwen3-Embedding-0.6B",
>     model_kwargs={
>         "gguf_file": '/mnt/Shelf/hf_home/Qwen3-Embedding-0.6B-Q8_0.gguf'
>     }
> )
> 
> This is probably too sketchy to be included in sentence transformers, but maybe this can be documented somewhere so that it's easier for people who need to run quants to find the instructions.
> 
> EDIT: you could probably also just place the `config.json` from the upstream hf repo into the same dir as the gguf, and use that directory as `model_name`. Haven't tested this though.


This doesn't (yet) work with embeddinggemma-300M QAT models, such as [this one](https://huggingface.co/unsloth/embeddinggemma-300m-GGUF).

While I haven't looked into this further, I'm curious if anyone knows if this is something `sentence-transformers` needs to update, or is this project waiting on dependencies to update for embeddinggemma as `sentence-transformers` asked me to install `torch` and `accelerate`.

```
model = SentenceTransformer(
    "google/embeddinggemma-300m",
    model_kwargs={
        "gguf_file": '/home/models/embeddinggemma-300M-Q8_0.gguf'
    }
)

# ValueError: GGUF model with architecture gemma-embedding is not supported yet.
```
